revocation/revocation.go

```diff
@@ -101,49 +110,142 @@ func NewWithOptions(opts Options) (Validator, error) {
 		opts.OCSPHTTPClient = &http.Client{Timeout: 2 * time.Second}
 	}
 
+	if opts.CRLHTTPClient == nil {
+		opts.CRLHTTPClient = &http.Client{Timeout: 5 * time.Second}
+	}
+
 	switch opts.CertChainPurpose {
 	case purpose.CodeSigning, purpose.Timestamping:
 	default:
 		return nil, fmt.Errorf("unsupported certificate chain purpose %v", opts.CertChainPurpose)
 	}
 
 	return &revocation{
-		httpClient:       opts.OCSPHTTPClient,
+		ocspHTTPClient:   opts.OCSPHTTPClient,
+		crlHTTPClient:    opts.CRLHTTPClient,
 		certChainPurpose: opts.CertChainPurpose,
 	}, nil
 }
 
 // Validate checks the revocation status for a certificate chain using OCSP and
-// returns an array of CertRevocationResults that contain the results and any
-// errors that are encountered during the process
+// CRL if OCSP is not available. It returns an array of CertRevocationResults
+// that contain the results and any errors that are encountered during the
+// process.
 //
-// TODO: add CRL support
-// https://github.com/notaryproject/notation-core-go/issues/125
+// This function tries OCSP and falls back to CRL when:
+// - OCSP is not supported by the certificate
+// - OCSP returns an unknown status
+//
+// NOTE: The certificate chain is expected to be in the order of leaf to root.
 func (r *revocation) Validate(certChain []*x509.Certificate, signingTime time.Time) ([]*result.CertRevocationResult, error) {
 	return r.ValidateContext(context.Background(), ValidateContextOptions{
 		CertChain:            certChain,
 		AuthenticSigningTime: signingTime,
 	})
 }
 
-// ValidateContext checks the revocation status for a certificate chain using
-// OCSP and returns an array of CertRevocationResults that contain the results
-// and any errors that are encountered during the process
+// ValidateContext checks the revocation status for a certificate chain using OCSP and
+// CRL if OCSP is not available. It returns an array of CertRevocationResults
+// that contain the results and any errors that are encountered during the
+// process.
+//
+// This function tries OCSP and falls back to CRL when:
+// - OCSP is not supported by the certificate
+// - OCSP returns an unknown status
 //
-// TODO: add CRL support
-// https://github.com/notaryproject/notation-core-go/issues/125
+// NOTE: The certificate chain is expected to be in the order of leaf to root.
 func (r *revocation) ValidateContext(ctx context.Context, validateContextOpts ValidateContextOptions) ([]*result.CertRevocationResult, error) {
```

Comment:
nit: What's the performance of this function? Can it fail fast? If so, please improve it in the next PR.

Reply:
Created an issue to track the performance optimization #221